### PR TITLE
CA1700: Do not name enum values 'Reserved'

### LINKED ISSUE #440: Port FxCop rule CA1700: DoNotNameEnumValuesReserved
**Title:** Do not name enum values 'Reserved'

**Description:**

This rule assumes that an enumeration member that has a name that contains ""reserved"" is not currently used but is a placeholder to be renamed or removed in a future version. Renaming or removing a member is a breaking change.

**Dependency:** None

**Notes:**

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #440

### CODE REVIEW

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/MicrosoftCodeQualityAnalyzersResources.resx [hunk 1]
```diff
@@ -1370,4 +1370,13 @@
   <data name="EnumShouldNotHaveDuplicatedValuesTitle" xml:space="preserve">
     <value>Enums values should not be duplicated</value>
   </data>
+  <data name="DoNotNameEnumValuesReservedDescription" xml:space="preserve">
+    <value>This rule assumes that an enumeration member that has a name that contains "reserved" is not currently used but is a placeholder to be renamed or removed in a future version. Renaming or removing a member is a breaking change.</value>
+  </data>
+  <data name="DoNotNameEnumValuesReservedMessage" xml:space="preserve">
+    <value>If '{0}.{1}' is not used in the current implementation, remove it. Otherwise give it a meaningful name.</value>
```
Comment (Evangelink) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/MicrosoftCodeQualityAnalyzersResources.resx]: Message is copied from legacy fxcop

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/MicrosoftCodeQualityAnalyzersResources.resx [hunk 2]
```diff
@@ -1370,4 +1370,13 @@
   <data name="EnumShouldNotHaveDuplicatedValuesTitle" xml:space="preserve">
     <value>Enums values should not be duplicated</value>
   </data>
+  <data name="DoNotNameEnumValuesReservedDescription" xml:space="preserve">
+    <value>This rule assumes that an enumeration member that has a name that contains "reserved" is not currently used but is a placeholder to be renamed or removed in a future version. Renaming or removing a member is a breaking change.</value>
```
Comment (Evangelink) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/MicrosoftCodeQualityAnalyzersResources.resx]: Copied from first part of the rule description.

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotNameEnumValuesReserved.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        private static readonly LocalizableString s_localizableMessageRule = new LocalizableResourceString(nameof(MicrosoftCodeQualityAnalyzersResources.DoNotNameEnumValuesReservedMessage), MicrosoftCodeQualityAnalyzersResources.ResourceManager, typeof(MicrosoftCodeQualityAnalyzersResources));
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(MicrosoftCodeQualityAnalyzersResources.DoNotNameEnumValuesReservedDescription), MicrosoftCodeQualityAnalyzersResources.ResourceManager, typeof(MicrosoftCodeQualityAnalyzersResources));
+
+        public static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(RuleId,
+                                                                                    s_localizableTitle,
+                                                                                    s_localizableMessageRule,
+                                                                                    DiagnosticCategory.Naming,
+                                                                                    RuleLevel.IdeHidden_BulkConfigurable,
```
Comment (Evangelink) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotNameEnumValuesReserved.cs]: What shall be the default level?

**Comment by mavasani** [comment]
FYI: You will need to fix a new RS diagnostic once https://github.com/dotnet/roslyn-analyzers/pull/3444 goes in.

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotNameEnumValuesReserved.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        public override void Initialize(AnalysisContext analysisContext)
+        {
+            analysisContext.EnableConcurrentExecution();
+            analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            analysisContext.RegisterSymbolAction(context =>
+            {
+                var field = (IFieldSymbol)context.Symbol;
+
+                // FxCop compat: only analyze externally visible symbols by default.
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotNameEnumValuesReserved.cs:41]: Thanks for handling this!

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotNameEnumValuesReserved.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+                if (!field.MatchesConfiguredVisibility(context.Options, Rule, context.CancellationToken))
+                {
+                    return;
+                }
+
+
+                if (field.ContainingType?.TypeKind == TypeKind.Enum &&
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotNameEnumValuesReserved.cs:48]: Nit: Might be cheaper to do this checks before the `MatchesConfiguredVisibility` check...

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotNameEnumValuesReserved.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+                    field.Name.Contains("Reserved"))
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotNameEnumValuesReserved.cs:49]: FxCop rule implementation uses `WordParser`, which has been ported to this repo, so we should also match it:

```csharp
if (!WordParser.ContainsWord(member.Name.Name, WordParserOptions.SplitCompoundWords, "reserved"))
           return null;
```

Comment (mavasani) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotNameEnumValuesReserved.cs:49]: You can likely write a test by using a field name such as `Preserved` or such.

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotNameEnumValuesReserved.cs [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+                if (field.ContainingType == null ||
+                    field.ContainingType.TypeKind != TypeKind.Enum ||
+                    !WordParser.ContainsWord(field.Name, WordParserOptions.SplitCompoundWords, reservedWords))
+                {
+                    return;
+                }
+
+                // FxCop compat: only analyze externally visible symbols by default.
+                if (!field.MatchesConfiguredVisibility(context.Options, Rule, context.CancellationToken))
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotNameEnumValuesReserved.cs:51]: Please update the analyzer configuration document that this rule supports `api_surface`.

**Pull request merged**